File: pyTools/features/trajectory/Burgos.py
import numpy as np


def generateTrajFeatures(traj, dt=0.033):
    """ Calculate Trajectory Features as in Burgos-Artizzu CVPR 2012

    Args:
        traj (ndarray):
                            two dimensional matrix with x,y
                            positions for each time step t
        dt (float):
                            time difference between two positions
                            in traj

    Returns:
        feat (ndarray):
                            N x 7 matrix
    """

    # convenience functions to zero pad t offsets
    # t -1
    t_m1 = lambda x: np.pad(x[:-1], (1, 0), mode='constant')
    # t - 2
    t_m2 = lambda x: np.pad(x[:-2], (2, 0), mode='constant')
    # t + 1
    t_p1 = lambda x: np.pad(x[1:], (0, 1), mode='constant')

    # position
    position = traj

    # traj differences
    dx = traj[:, 0] - t_m1(traj[:, 0])
    dy = traj[:, 1] - t_m1(traj[:, 1])

    # direction
    # direction = dy / dx
    # direction[np.isnan(direction)] = 0
    direction = np.arctan2(dy, dx)

    dir_change = direction - t_m1(direction)

    # velocity
    vx = 1 / dt * (0.25 * t_m1(dx) + 0.5 * dx + 0.25 * t_p1(dx))
    vy = 1 / dt * (0.25 * t_m1(dy) + 0.5 * dy + 0.25 * t_p1(dy))

    # acceleration
    ax = (t_p1(vx) - t_m1(vx)) / (2 * dt)
    ay = (t_p1(vy) - t_m1(vy)) / (2 * dt)

    # concatenate feature vector/matrix
    # using np.c_ instead of np.concatenate to avoid problems with
    # shape of position and the other features
    feat = np.c_[position, direction, dir_change,
                           vx, vy, ax, ay]

    return feat

File: pyTools/features/trajectory/Burgos.m
function [ feats ] = generateTrajFeatures( traj )
% traj is 2D matrix numFrames*[X Y]
% feats is a numFrames * 8 matrix
% feats is the feature matrix from Burgos-Artizzu CVPR 2012 - top Pg 6

% TODOs
% not sure about deltaT or atan
% brittle - very susceptible to bad trajectories and NaNs
% extend to two object features

% position
f1 = [traj];

% direction
f2_1 = (traj(:,2) - [0; traj(1:end-1,2)]) ./ (traj(:,1) - [0; traj(1:end-1,1)]);
f2_1(isnan(f2_1)) = 0; % correct?
f2_1 = atan(f2_1); %correct?

% dir change
f3_1 = f2_1 - [0; f2_1(1:end-1)];

% velocity
x1_d = (traj(:,1) - [0; traj(1:end-1,1)]);
y1_d = (traj(:,2) - [0; traj(1:end-1,2)]);
deltaT = (1/3); % ?? not sure what this is supposed to be

f4_x1 = deltaT*(0.25*[0; x1_d(1:end-1,:)] + 0.5*x1_d +0.25*[x1_d(2:end,:); 0]);
f4_y1 = deltaT*(0.25*[0; y1_d(1:end-1,:)] + 0.5*y1_d +0.25*[y1_d(2:end,:); 0]);

% acceleration
deltaT2 = 1/(2*2); % ?? not sure what this is
f5_x1 = ([f4_x1(2:end) ; 0] - [0; f4_x1(1:end-1)])*deltaT2;
f5_y1 = ([f4_y1(2:end) ; 0] - [0; f4_y1(1:end-1)])*deltaT2;

feats = [f1, f2_1, f3_1, f4_x1, f4_y1, f5_x1, f5_y1,]; 